### PR TITLE
Add support for PKCS#1 keys

### PR DESCRIPTION
**Pull request by bgregos**
A quick and dirty implementation that allows for PKCS#1 keys. This is helpful for taskservers like inthe.am. The major downside to adding this code is that BouncyCastle is now added as a dependency, but I don't at present see a good way around it.

### CODE REVIEW

**Comment by bgregos** [comment]
I see a good bit of low-hanging fruit for cleanup, I'll try to update this soon.

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 1]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
-        byte[] privKeyBytes;
+    public PrivateKey createPrivateKeyFor(File privateKeyFile) {
+        byte[] privKeyBytes = null;
         try (BufferedInputStream bis = new BufferedInputStream(new FileInputStream(privateKeyFile))) {
             privKeyBytes = new byte[(int) privateKeyFile.length()];
             bis.read(privKeyBytes);
         } catch (IOException e) {
-            throw new TaskwarriorKeyStoreException(e, "Could not read private key '%s' via input stream: %s", privateKeyFile,
-                    e.getMessage());
+            e.printStackTrace();
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:142]: Why removed throwing the exception?

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 2]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
         }
 
+        String privateKeyString = "----...";
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:145]: The rest should go to a separate method.

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 3]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
         try {
-            KeyFactory keyFactory = KeyFactory.getInstance(ALGORITHM_PRIVATE_KEY);
-            return keyFactory.generatePrivate(new PKCS8EncodedKeySpec(privKeyBytes));
-        } catch (NoSuchAlgorithmException e) {
-            throw new TaskwarriorKeyStoreException(e, "Key factory could not be initialized for algorithm '%s': %s", ALGORITHM_PRIVATE_KEY,
-                    e.getMessage());
-        } catch (InvalidKeySpecException e) {
-            throw new TaskwarriorKeyStoreException(e, "Could not generate private key for '%s': %s", privateKeyFile, e.getMessage());
+            PemReader pemReader;
+            pemReader = new PemReader(new InputStreamReader(new ByteArrayInputStream(privKeyBytes)));
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:150]: Merge with the line above.

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 4]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
+
+            privateKeyObject = pemReader.readPemObject();
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:154]: I prefer throwing an exception or logging it. What's your opinion?

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 5]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
+            //PKCS#8 key
+            try {
+                privateKeyParameter = (RSAPrivateCrtKeyParameters) PrivateKeyFactory.createKey(
+                        privateKeyObject.getContent()
+                );
+            } catch (IOException e) {
+                e.printStackTrace();
+            }
+        } else {
+            throw new RuntimeException("Unsupported key type: " + privateKeyObject.getType());
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:181]: `UnsupportedOperationException`?

File: src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java [hunk 6]
```diff
@@ -123,24 +133,64 @@ public KeyStore build() {
...
+
+        try {
+            return new JcaPEMKeyConverter()
+                    .getPrivateKey(
+                            PrivateKeyInfoFactory.createPrivateKeyInfo(
+                                    privateKeyParameter
+                            )
+                    );
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (aaschmid) [inline src/main/java/de/aaschmid/taskwarrior/ssl/KeyStoreBuilder.java:192]: ;-)

**Comment by aaschmid** [comment]
@bgregos: Thanks for your changes. In sum I like them pretty much. I would appreciate some tests, though. But as I haven't had some myself, I thing I first need to provide some infrastructure for testing this.
Can you discuss my review points?

File: build.gradle [hunk 7]
```diff
@@ -20,6 +20,11 @@ task wrapper(type: Wrapper) {
     gradleVersion = '4.6'
 }
 
+dependencies {
+    // https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15on
+    implementation group: 'org.bouncycastle', name: 'bcprov-jdk15on', version: '1.60'
```
Comment (aaschmid) [inline build.gradle]: Would it also be possible to use very much smaller `bcpkix-jdk15on` artifact?

**Review by aaschmid — CHANGES_REQUESTED** [review]
I am still not sure if I want to include such a "huge"(TM) library into my "core" application but maybe I will provide another version. There is one other possibility (see http://magnus-k-karlsson.blogspot.com/2018/05/how-to-read-pem-pkcs1-or-pkcs8-encoded.html) with the ugly tradeoff that it uses `sun.*` packages. Would they be available on android jdk?

File: build.gradle (hunk 7 above)
Comment (aaschmid) [inline build.gradle:25]: I prefer to create a shadow jar including BC using https://imperceptiblethoughts.com/shadow/. But we could that later on as well...

**Comment by aaschmid** [comment]
And it would be nice to sign-off your commits as stated in https://github.com/aaschmid/taskwarrior-java-client/blob/master/CONTRIBUTING.md and agree the terms of the license agreement. Thanks @bgregos

File: build.gradle [hunk 8]
```diff
@@ -23,12 +23,8 @@ task wrapper(type: Wrapper) {
 dependencies {
     // https://mvnrepository.com/artifact/org.bouncycastle/bcprov-jdk15on
     implementation group: 'org.bouncycastle', name: 'bcprov-jdk15on', version: '1.60'
```
Comment (aaschmid) [inline build.gradle:25]: I need to use `org.bouncycastle:bcpkix-jdk15on` to satisfy all used classes in your code.

**Comment by aaschmid** [comment]
Maybe you can already rebase your changes onto https://github.com/aaschmid/taskwarrior-java-client/tree/pkcs1.and.pem.support.%232. That would be very helpful @bgregos. But please give me a hint if I should do myself :-)

**Comment by aaschmid** [comment]
@bgregos: I merged your branch into a branch of mine in order to use your changes to my code for a version use bouncycastle for support of PKCS#1 keys. I would be very nice if you could agree the terms of the license agreement as in https://github.com/aaschmid/taskwarrior-java-client/blob/master/CONTRIBUTING.md. Thanks!

**Comment by bgregos** [comment]
Hello, apologies for the delay. I hereby agree to the terms of the Taskwarrior Java client License Agreement. Thanks for your additions to the PR as well; let me know if anything else needs done in relation to this PR.

**Comment by aaschmid** [comment]
Thanks, all the things are included in the version on Maven Central. 

I am actually curious about your results including the published version into your Android app :-)

**Pull request merged**